File: leetcode/098_validate_binary_search_tree.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <cassert>
#include <cstddef>
#include <cstdint>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;

    explicit TreeNode(int x): val(x), left(NULL), right(NULL) {}
};

class Solution {
 public:
    bool isValidBST(TreeNode* root) {
        return isValidBSTHelper(root, INT64_MIN, INT64_MAX);
    }

 private:
    bool isValidBSTHelper(TreeNode* root, int64_t minimum, int64_t maximum) {
        if (!root) {
            return true;
        } else {
            return minimum < root->val && root->val < maximum &&
              isValidBSTHelper(root->left, minimum, root->val) &&
              isValidBSTHelper(root->right, root->val, maximum);
        }
    }
};

int main() {
    auto solution = Solution();

    auto t0_0 = TreeNode(2);
    auto t0_1 = TreeNode(1);
    auto t0_2 = TreeNode(3);
    t0_0.right = &t0_2;
    t0_0.left = &t0_1;

    assert(solution.isValidBST(&t0_0));

    auto t1_0 = TreeNode(5);
    auto t1_1 = TreeNode(1);
    auto t1_2 = TreeNode(4);
    auto t1_3 = TreeNode(3);
    auto t1_4 = TreeNode(6);
    t1_2.right = &t1_4;
    t1_2.left = &t1_3;
    t1_0.right = &t1_2;
    t1_0.left = &t1_1;

    assert(!solution.isValidBST(&t1_0));

    auto t2_0 = TreeNode(2147483647);

    assert(solution.isValidBST(&t2_0));
}

File: leetcode/098_validate_binary_search_tree.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def isValidBST(self, root):
        return self._isValidBST(root, float('-inf'), float('inf'))

    def _isValidBST(self, root, minimum, maximum):
        if root is None:
            return True
        return (
            minimum < root.val < maximum and
            self._isValidBST(root.left, minimum, root.val) and
            self._isValidBST(root.right, root.val, maximum)
        )


if __name__ == '__main__':
    solution = Solution()

    t0_0 = TreeNode(2)
    t0_1 = TreeNode(1)
    t0_2 = TreeNode(3)
    t0_0.right = t0_2
    t0_0.left = t0_1

    assert solution.isValidBST(t0_0)

    t1_0 = TreeNode(5)
    t1_1 = TreeNode(1)
    t1_2 = TreeNode(4)
    t1_3 = TreeNode(3)
    t1_4 = TreeNode(6)
    t1_2.right = t1_4
    t1_2.left = t1_3
    t1_0.right = t1_2
    t1_0.left = t1_1

    assert not solution.isValidBST(t1_0)

    t2_0 = TreeNode(2147483647)

    assert solution.isValidBST(t2_0)
